File: Packages/MeterKit/Sources/MeterFeatures/Settings/TipModel.swift
import Foundation
import Observation
import SwiftData
import MeterPersistence
import MeterTips

@MainActor
@Observable
final class TipModel {
    enum CatalogState: Equatable {
        case loading
        case ready
        case unavailable
    }

    enum PurchaseNotice: Equatable {
        case cancelled
        case pending
        case failed
    }

    var catalogState: CatalogState = .loading
    var offerings: [TipOffering] = []
    var records: [TipHistoryItem] = []
    var isPurchasing = false
    var purchaseNotice: PurchaseNotice?
    var thanksVisible = false
    /// 这次买的是哪一档 / 是不是回头客——付款成功那两句照这两个值挑，见 `TipThanks`。
    var thanksTreat: TipProductID?
    var thanksIsRepeat = false
    var composerVisible = false
    var draftName = ""
    var draftMessage = ""
    var pendingTransactionID: String?
    var willRetryMessage = false

    private let container: ModelContainer
    private let catalog: any TipCataloging
    private let purchaser: any TipPurchasing
    private let messenger: any TipMessageSubmitting
    private let listener: any TipTransactionListening
    private let appVersion: String

    init(
        container: ModelContainer,
        catalog: any TipCataloging = StoreKitTipCatalog(),
        purchaser: any TipPurchasing = StoreKitTipPurchaser(),
        messenger: any TipMessageSubmitting = TipWorkerClient(),
        listener: any TipTransactionListening = StoreKitTipTransactionListener(),
        appVersion: String = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.1.0"
    ) {
        self.container = container
        self.catalog = catalog
        self.purchaser = purchaser
        self.messenger = messenger
        self.listener = listener
        self.appVersion = appVersion
        reloadHistory()
        thanksVisible = !records.isEmpty
    }

    var thanks: TipThanks {
        TipThanks.make(treat: thanksTreat, isRepeat: thanksIsRepeat)
    }

    func load() async {
        catalogState = .loading
        do {
            var products = try await catalog.loadProducts()
            if products.isEmpty {
                // StoreKit 本地店面刚挂上时，第一次查询经常是空数组。
                try await Task.sleep(for: .milliseconds(400))
                products = try await catalog.loadProducts()
            }
            if products.isEmpty {
                offerings = []
                catalogState = .unavailable
            } else {
                offerings = products
                catalogState = .ready
            }
        } catch {
            offerings = []
            catalogState = .unavailable
        }
        reloadHistory()
        await retryUnsentMessages()
    }

    func observeTransactions() async {
        for await transaction in listener.updates() {
            ingest(transaction, showComposer: true)
            await retryUnsentMessages()
        }
    }

    func buy(_ offering: TipOffering) async {
        guard !isPurchasing else { return }
        isPurchasing = true
        purchaseNotice = nil
        let outcome = await purchaser.purchase(productID: offering.id)
        isPurchasing = false

        switch outcome {
        case .success(let transaction):
            ingest(transaction, showComposer: true)
        case .cancelled:
            purchaseNotice = .cancelled
        case .pending:
            purchaseNotice = .pending
        case .failed:
            purchaseNotice = .failed
        }
    }

    func submitComposer() async {
        guard let transactionID = pendingTransactionID else {
            composerVisible = false
            return
        }
        applyDraft(to: transactionID)
        composerVisible = false
        pendingTransactionID = nil
        await submit(transactionID: transactionID)
    }

    func skipComposer() async {
        guard let transactionID = pendingTransactionID else {
            composerVisible = false
            return
        }
        composerVisible = false
        pendingTransactionID = nil
        await submit(transactionID: transactionID)
    }

    /// 划掉这页等于「这次不留」。已经点过写好了就什么都不做。
    func abandonComposerIfNeeded() async {
        guard composerVisible, pendingTransactionID != nil else { return }
        await skipComposer()
    }

    func retryUnsentMessages() async {
        let context = ModelContext(container)
        let pending = (try? TipRecord.unsynced(from: context)) ?? []
        for record in pending {
            if composerVisible, record.transactionID == pendingTransactionID {
                continue
            }
            await submit(transactionID: record.transactionID)
        }
    }

    private func ingest(_ transaction: TipTransaction, showComposer: Bool) {
        let context = ModelContext(container)
        let existed = (try? TipRecord.fetch(
            transactionID: transaction.id,
            from: context
        )) != nil
        _ = try? TipRecord.upsert(
            transactionID: transaction.id,
            productID: transaction.productID,
            displayPrice: transaction.displayPrice,
            purchasedAt: transaction.purchasedAt,
            jws: transaction.jws,
            appVersion: appVersion,
            in: context
        )
        reloadHistory()
        thanksVisible = true
        thanksTreat = TipProductID(rawValue: transaction.productID)
        // reloadHistory 已经把这一笔算进去了，所以「不止一笔」就是回头客。
        thanksIsRepeat = records.count > 1
        purchaseNotice = nil
        if showComposer, !existed {
            pendingTransactionID = transaction.id
            composerVisible = true
            draftName = ""
            draftMessage = ""
        }
    }

    private func applyDraft(to transactionID: String) {
        let context = ModelContext(container)
        guard let record = try? TipRecord.fetch(transactionID: transactionID, from: context) else {
            return
        }
        record.name = TipFieldLimits.clampName(draftName)
        record.message = TipFieldLimits.clampMessage(draftMessage)
        try? context.save()
        reloadHistory()
    }

    private func submit(transactionID: String) async {
        let context = ModelContext(container)
        guard let record = try? TipRecord.fetch(transactionID: transactionID, from: context) else {
            return
        }
        let payload = TipMessagePayload(
            transactionID: record.transactionID,
            jws: record.jws,
            productID: record.productID,
            displayPrice: record.displayPrice,
            name: record.name,
            message: record.message,
            appVersion: record.appVersion
        )
        do {
            try await messenger.submit(payload)
            let latest = ModelContext(container)
            if let stored = try TipRecord.fetch(transactionID: transactionID, from: latest) {
                stored.isSubmitted = true
                try latest.save()
            }
            let remaining = (try? TipRecord.unsynced(from: ModelContext(container))) ?? []
            willRetryMessage = remaining.contains { $0.transactionID != pendingTransactionID }
            reloadHistory()
        } catch {
            willRetryMessage = true
            reloadHistory()
        }
    }

    private func reloadHistory() {
        let context = ModelContext(container)
        let stored = (try? TipRecord.all(from: context)) ?? []
        records = stored.map { record in
            TipHistoryItem(
                transactionID: record.transactionID,
                productID: record.productID,
                displayPrice: record.displayPrice,
                purchasedAt: record.purchasedAt,
                name: record.name,
                message: record.message,
                isSubmitted: record.isSubmitted
            )
        }
    }
}

extension TipModel {
    static var preview: TipModel {
        preview(seedHistory: false, catalog: .loaded)
    }

    static var previewUnavailable: TipModel {
        preview(seedHistory: false, catalog: .unavailable)
    }

    static var previewHistory: TipModel {
        preview(seedHistory: true, catalog: .loaded)
    }

    static var previewComposer: TipModel {
        let model = preview(seedHistory: true, catalog: .loaded)
        model.composerVisible = true
        model.pendingTransactionID = "preview-1"
        model.thanksTreat = .large
        return model
    }

    private enum PreviewCatalog {
        case loaded
        case unavailable
    }

    private static func preview(seedHistory: Bool, catalog: PreviewCatalog) -> TipModel {
        let container = try! PersistenceContainer.makeContainer(inMemory: true)
        if seedHistory {
            let context = ModelContext(container)
            _ = try? TipRecord.upsert(
                transactionID: "preview-1",
                productID: TipProductID.small.rawValue,
                displayPrice: "6.00",
                purchasedAt: Date(timeIntervalSince1970: 1_787_000_000),
                name: String(localized: L("陈")),
                message: String(localized: L("好用")),
                jws: "preview-jws",
                appVersion: "0.1.0",
                in: context
            )
        }
        let model = TipModel(
            container: container,
            catalog: PreviewTipCatalog(isUnavailable: catalog == .unavailable),
            purchaser: PreviewTipPurchaser(),
            messenger: PreviewTipMessenger(),
            listener: IdleTipTransactionListener(),
            appVersion: "0.1.0"
        )
        if seedHistory {
            model.thanksVisible = true
        }
        return model
    }
}

private struct PreviewTipCatalog: TipCataloging {
    var isUnavailable: Bool

    func loadProducts() async throws -> [TipOffering] {
        if isUnavailable { return [] }
        return [
            TipOffering(id: TipProductID.small.rawValue, displayName: TipProductID.small.listTitle, displayPrice: "6.00"),
            TipOffering(id: TipProductID.medium.rawValue, displayName: TipProductID.medium.listTitle, displayPrice: "18.00"),
            TipOffering(id: TipProductID.large.rawValue, displayName: TipProductID.large.listTitle, displayPrice: "45.00"),
        ]
    }
}

private struct PreviewTipPurchaser: TipPurchasing {
    func purchase(productID: String) async -> TipPurchaseOutcome {
        .success(
            TipTransaction(
                id: "preview-\(productID)",
                productID: productID,
                displayPrice: "6.00",
                jws: "preview-jws",
                purchasedAt: Date(timeIntervalSince1970: 1_787_000_000)
            )
        )
    }
}

private struct PreviewTipMessenger: TipMessageSubmitting {
    func submit(_ payload: TipMessagePayload) async throws {}
}

private struct IdleTipTransactionListener: TipTransactionListening {
    func updates() -> AsyncStream<TipTransaction> {
        AsyncStream { continuation in
            continuation.finish()
        }
    }
}

File: Android/app/src/main/kotlin/com/zhechengqi/tollcat/settings/TipModel.kt
package com.zhechengqi.tollcat.settings

import android.app.Activity
import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.android.billingclient.api.BillingClient
import com.zhechengqi.tollcat.R
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class TipModel(
    context: Context,
    private val store: TipStore = TipStore(context),
    private val billing: PlayBillingTips? = PlayBillingTips(context),
    private val messenger: (TipMessagePayload) -> Boolean = { TipMessageClient.submit(it) },
    private val appVersion: String = appVersionShort(context),
    previewOfferings: List<TipOffering>? = null,
) {
    enum class CatalogState { Loading, Ready, Unavailable }

    var catalogState by mutableStateOf(if (previewOfferings == null) CatalogState.Loading else CatalogState.Ready)
        private set
    var offerings by mutableStateOf(previewOfferings.orEmpty())
        private set
    var records by mutableStateOf(store.all())
        private set
    var isPurchasing by mutableStateOf(false)
        private set
    var purchaseNotice by mutableStateOf<TipPurchaseNotice?>(null)
        private set
    var thanksVisible by mutableStateOf(records.isNotEmpty())
        private set
    var thanksTreat by mutableStateOf<TipProductID?>(null)
        private set
    var thanksIsRepeat by mutableStateOf(false)
        private set
    var composerVisible by mutableStateOf(false)
        private set
    var draftName by mutableStateOf("")
    var draftMessage by mutableStateOf("")
    var pendingTransactionID by mutableStateOf<String?>(null)
        private set
    var willRetryMessage by mutableStateOf(false)
        private set

    val thanks: TipThanks get() = TipThanks.make(thanksTreat, thanksIsRepeat)

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val preview = previewOfferings != null

    init {
        billing?.events = object : PlayBillingTips.Events {
            override fun onTransaction(transaction: TipTransaction) {
                ingest(transaction, showComposer = true)
                isPurchasing = false
                scope.launch { billing.consume(transaction.id) }
            }

            override fun onCancelled() {
                isPurchasing = false
                purchaseNotice = TipPurchaseNotice.Cancelled
            }

            override fun onPending() {
                isPurchasing = false
                purchaseNotice = TipPurchaseNotice.Pending
            }

            override fun onFailed() {
                isPurchasing = false
                purchaseNotice = TipPurchaseNotice.Failed
            }
        }
    }

    fun start() {
        if (preview) {
            scope.launch { retryUnsentMessages() }
            return
        }
        scope.launch {
            catalogState = CatalogState.Loading
            billing?.start()
            val products = billing?.loadProducts().orEmpty()
            if (products.isEmpty()) {
                offerings = emptyList()
                catalogState = CatalogState.Unavailable
            } else {
                offerings = products
                catalogState = CatalogState.Ready
            }
            reloadHistory()
            billing?.restoreUnconsumed()
            retryUnsentMessages()
        }
    }

    fun release() {
        scope.cancel()
        billing?.end()
    }

    fun buy(activity: Activity, offering: TipOffering) {
        if (isPurchasing) return
        isPurchasing = true
        purchaseNotice = null
        if (preview) {
            ingest(
                TipTransaction(
                    id = "preview-${offering.id}-${System.currentTimeMillis()}",
                    productID = offering.id,
                    displayPrice = offering.displayPrice,
                    jws = "preview",
                    purchasedAtMillis = System.currentTimeMillis(),
                ),
                showComposer = true,
            )
            isPurchasing = false
            return
        }
        val code = billing?.launchPurchase(activity, offering.id)
            ?: BillingClient.BillingResponseCode.ERROR
        when (code) {
            BillingClient.BillingResponseCode.OK -> Unit
            BillingClient.BillingResponseCode.ITEM_ALREADY_OWNED -> {
                isPurchasing = false
                scope.launch { billing?.restoreUnconsumed() }
            }
            BillingClient.BillingResponseCode.USER_CANCELED -> {
                isPurchasing = false
                purchaseNotice = TipPurchaseNotice.Cancelled
            }
            else -> {
                isPurchasing = false
                purchaseNotice = TipPurchaseNotice.Failed
            }
        }
    }

    fun submitComposer() {
        val transactionID = pendingTransactionID
        if (transactionID == null) {
            composerVisible = false
            return
        }
        store.updateDraft(
            transactionID,
            TipFieldLimits.clampName(draftName),
            TipFieldLimits.clampMessage(draftMessage),
        )
        composerVisible = false
        pendingTransactionID = null
        scope.launch { submit(transactionID) }
    }

    fun skipComposer() {
        val transactionID = pendingTransactionID
        if (transactionID == null) {
            composerVisible = false
            return
        }
        composerVisible = false
        pendingTransactionID = null
        scope.launch { submit(transactionID) }
    }

    /** 划掉这页等于「这次不留」。已经点过写好了就什么都不做。 */
    fun abandonComposerIfNeeded() {
        if (composerVisible && pendingTransactionID != null) skipComposer()
    }

    private suspend fun retryUnsentMessages() {
        val pending = store.unsynced()
        for (record in pending) {
            if (composerVisible && record.transactionID == pendingTransactionID) continue
            submit(record.transactionID)
        }
    }

    private fun ingest(transaction: TipTransaction, showComposer: Boolean) {
        val (_, created) = store.upsert(
            transactionID = transaction.id,
            productID = transaction.productID,
            displayPrice = transaction.displayPrice,
            purchasedAtMillis = transaction.purchasedAtMillis,
            jws = transaction.jws,
            appVersion = appVersion,
        )
        reloadHistory()
        thanksVisible = true
        thanksTreat = TipProductID.fromRaw(transaction.productID)
        thanksIsRepeat = records.size > 1
        purchaseNotice = null
        if (showComposer && created) {
            pendingTransactionID = transaction.id
            composerVisible = true
            draftName = ""
            draftMessage = ""
        }
    }

    private suspend fun submit(transactionID: String) {
        val record = store.fetch(transactionID) ?: return
        val payload = TipMessagePayload.of(
            transactionID = record.transactionID,
            jws = record.jws,
            productID = record.productID,
            displayPrice = record.displayPrice,
            name = record.name,
            message = record.message,
            appVersion = record.appVersion,
        )
        val ok = withContext(Dispatchers.IO) { messenger(payload) }
        if (ok) {
            store.markSubmitted(transactionID)
            val remaining = store.unsynced()
            willRetryMessage = remaining.any { it.transactionID != pendingTransactionID }
            reloadHistory()
        } else {
            willRetryMessage = true
            reloadHistory()
        }
    }

    private fun reloadHistory() {
        records = store.all()
    }

    companion object {
        fun preview(context: Context, seedHistory: Boolean = false, unavailable: Boolean = false): TipModel {
            val store = TipStore(context)
            if (seedHistory && store.fetch("preview-1") == null) {
                store.upsert(
                    transactionID = "preview-1",
                    productID = TipProductID.Small.raw,
                    displayPrice = "6.00",
                    purchasedAtMillis = 1_787_000_000_000L,
                    name = "陈",
                    message = "好用",
                    jws = "preview-jws",
                    appVersion = "0.1.0",
                )
            }
            val offerings = if (unavailable) {
                emptyList()
            } else {
                listOf(
                    TipOffering(TipProductID.Small.raw, context.getString(R.string.settings_tip_candy), "¥6.00"),
                    TipOffering(TipProductID.Medium.raw, context.getString(R.string.settings_tip_coffee), "¥18.00"),
                    TipOffering(TipProductID.Large.raw, context.getString(R.string.settings_tip_pizza), "¥45.00"),
                )
            }
            val model = TipModel(
                context = context,
                store = store,
                billing = null,
                messenger = { true },
                appVersion = "0.1.0",
                previewOfferings = offerings,
            )
            if (unavailable) model.catalogState = CatalogState.Unavailable
            if (seedHistory) model.thanksVisible = true
            return model
        }
    }

    internal fun previewOpenComposer() {
        composerVisible = true
        pendingTransactionID = records.firstOrNull()?.transactionID ?: "preview-1"
        thanksTreat = TipProductID.Large
        thanksVisible = true
    }
}
